R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- setdiff(class(x), c(oldClass, "environment"))
+  ret
+}
+
+
+# Do not export to encourage direct use of `tagQuery()`
+asTagEnv <- function(x) {
+  if (isTagQuery(x)) {
+    return(asTagEnv(x$root()))
+  }
+
+  if (!isTagEnv(x)) {
+    if (!isTag(x)) {
+      # force all methods to send in tags, lists / tagLists are not allowed
+      stop("`asTagEnv()` can only accept tag envs or tag objects. It does not accept `lists()` or `tagLists()`")
+    }
+  }
+  asTagEnv_(x, parent = x$parent)
+}
+asTagEnv_ <- function(x, parent = NULL, seenMap = envirMap()) {
+  isTagVal <- isTag(x)
+  isTagEnvVal <- isTagEnv(x)
+
+  if (isTagVal || isTagEnvVal) {
+    if (!isTagEnvVal) {
+      xList <- x
+      x <- safeListToEnv(xList, "htmltools.tag.env")
+      # add parent env and key
+      x$parent <- parent
+      x$envKey <- sexp_address(x)
+    }
+    if (seenMap$has(x)) {
+      stop(
+        "Circular family tree found with tag environment: ", sexp_address(x), "\n",
+        # Not necessarily the order of the circular dependency
+        # TODO-later show actual circular dependency and not all visited nodes? This should be rare
+        "Tags processed: (unordered set)\n", paste0("* ", seenMap$keys(), collapse = "\n")
+      )
+    }
+    # Add the item to the seen map to help with cycle detection
+    seenMap$set(x, TRUE)
+
+    # Make sure all attribs are unique
+    x$attribs <- flattenTagAttribs(x$attribs)
+
+    # Recurse through children
+    if (length(x$children) != 0) {
+      # Possible optimization... name the children tags to the formatted values.
+      # * Allows for faster child look up later.
+      # * Comes with the cost of always formatting the env values even if children names are not needed.
+      # Attributes may be dropped
+      # * Could replace with `x$children[] <- ....`
+      # * Leaving as is to see if people mis-use the children field
+      x$children <- unname(lapply(
+        # Simplify the structures by flatting the tags
+        # Does NOT recurse to grand-children etc.
+        flattenTagsRaw(x$children),
+        # recurse through each child
+        asTagEnv_,
+        parent = x,
+        seenMap = seenMap
+      ))
+    }
+    # Remove the item from the map to allow for checks for ciruclar deps
+    # Having multiple child objects that are the same is ok, as long as a cycle is not found
+    seenMap$remove(x)
+  }
+  x
+}
+
+# This method MUST undo everything done in `asTagEnv(x)`
+# Do not export to encourage direct use of `tagQuery()$asTags()`
+tagEnvToTags <- function(x) {
+  if (isTagEnv(x)) {
+    xEl <- x
+    # convert to list first to avoid altering the original env obj
+    x <- safeEnvToList(xEl, c("htmltools.tag.env"))
+    # undo parent env and key
+    x$parent <- NULL
+    x$envKey <- NULL
+    # recurse through children
+    if (!is.null(x$children)) {
```

Comment:
Is `x$children` ever NULL or not present?

Reply:
It is safe to always call the `lapply`. Removing the check